text/0000-promote-libc.md

```diff
@@ -0,0 +1,308 @@
+[backcompat]: https://github.com/rust-lang/rust/pull/27762
+
+In order to solve this problem altogether, the libc crate will be enhanced with
+the ability to automatically test the FFI bindings it contains. As this crate
+will begin to live in `rust-lang` instead of the `rust` repo itself, this means
+it can leverage external CI systems like Travis CI and AppVeyor to perform these
+tasks.
+
+The [current implementation][ctest] of the binding testing verifies attributes
+such as type size/alignment, struct field offset, struct field types, constant
+values, function definitions, etc. Over time it can be enhanced with more
+metrics and properties to test.
+
+[ctest]: https://github.com/alexcrichton/ctest
+
+In theory adding a new platform to `libc` will be blocked until automation can
+be set up to ensure that the bindings are correct, but it is unfortunately not
+easy to add this form of automation for all platforms, so this will not be a
+requirement (beyond "tier 1 platforms"). There is currently automation for the
+following targets, however, through Travis and AppVeyor:
+
+* `{i686,x86_64}-pc-windows-{msvc,gnu}`
+* `{i686,x86_64,mips,aarch64}-unknown-linux-gnu`
+* `x86_64-unknown-linux-musl`
+* `arm-unknown-linux-gnueabihf`
+* `arm-linux-androideabi`
+* `{i686,x86_64}-apple-{darwin,ios}`
+
+# Drawbacks
+
+### Loss of module organization
+
+The loss of an internal organization structure can be seen as a drawback of this
+design. While perhaps not precisely true today, the principle of the structure
+was that it is easy to constrain yourself to a particular C standard or subset
+of C to in theory write "more portable programs by default" by only using the
+contents of the respective module. Unfortunately in practice this does not seem
+to be that much in use, and it's also not clear whether this can be expressed
+through simply headers in `libc`. For example many platforms will have slight
+tweaks to common structures, definitions, or types in terms of signedness or
+value, so even if you were restricted to a particular subset it's not clear that
+a program would automatically be more portable.
+
+That being said, it would still be useful to have these abstractions to *some
+degree*, but the filp side is that it's easy to build this sort of layer on top
+of `libc` as designed here externally on crates.io. For example `extern crate
+posix` could just depend on `libc` and reexport all the contents for the
+POSIX standard, perhaps with tweaked signatures here and there to work better
+across platforms.
+
+### Loss of Windows bindings
+
+By only exposing the CRT functions on Windows, the contents of `libc` will be
+quite trimmed down which means when accessing similar functions like `send` or
+`connect` crates will be required to link to two libraries at least.
+
+This is also a bit of a maintenance burden on the standard library itself as it
+means that all the bindings it uses must move to `src/libstd/sys/windows/c.rs`
+in the immedidate future.
```

Comment:
Although with a bit of build system finagling (like say, making Rust use Cargo as its build system), you could make it use `winapi` for those things. :wink: